hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/HoodieWriteHandle.java

```diff
@@ -138,9 +139,35 @@ protected Path makeNewFilePath(String partitionPath, String fileName) {
    *
    * @param partitionPath Partition path
    */
-  protected void createMarkerFile(String partitionPath, String dataFileName) {
-    WriteMarkersFactory.get(config.getMarkersType(), hoodieTable, instantTime)
-        .create(partitionPath, dataFileName, getIOType(), config, fileId, hoodieTable.getMetaClient().getActiveTimeline());
+  protected void createInProgressMarkerFile(String partitionPath, String dataFileName, String markerInstantTime) {
+    WriteMarkers writeMarkers = WriteMarkersFactory.get(config.getMarkersType(), hoodieTable, instantTime);
+    if (!writeMarkers.doesMarkerDirExist()) {
+      throw new HoodieIOException(String.format("Marker root directory absent : %s/%s (%s)",
+          partitionPath, dataFileName, markerInstantTime));
+    }
+    if (config.enforceFinalizeWriteCheck()
+        && writeMarkers.markerExists(writeMarkers.getCompletionMarkerPath("", "FINALIZE_WRITE", markerInstantTime, IOType.CREATE))) {
+      throw new HoodieCorruptedDataException("Reconciliation for instant " + instantTime + " is completed, job is trying to re-write the data files.");
+    }
+    if (config.enforceCompletionMarkerCheck()
+        && writeMarkers.markerExists(writeMarkers.getCompletionMarkerPath(partitionPath, fileId, markerInstantTime, getIOType()))) {
+      throw new HoodieIOException("Completed marker file exists for : " + dataFileName + " (" + instantTime + ")");
+    }
+    writeMarkers.create(partitionPath, dataFileName, getIOType());
+  }
+
+  // visible for testing
+  public void createCompletedMarkerFile(String partition, String markerInstantTime) throws IOException {
+    try {
+      WriteMarkersFactory.get(config.getMarkersType(), hoodieTable, instantTime)
+          .createCompletionMarker(partition, fileId, markerInstantTime, getIOType(), true);
+    } catch (Exception e) {
+      // Clean up the data file, if the marker is already present or marker directories don't exist.
+      Path partitionPath = FSUtils.getPartitionPath(hoodieTable.getMetaClient().getBasePath(), partition);
```

Comment:
After the finalizeWrite and reconciling the files, we delete the markerDirectory.   If a stray executor were to complete the write operation and close the file after the reconcile step, it would find markerDirectory missing and would cleanup the datafile created.

![Screenshot 2023-06-25 at 9 15 35 PM](https://github.com/apache/hudi/assets/47542891/f84e70f9-5f17-4454-8ff1-608c59056ef3)
In the example, executor C trying to close the file, after finalizeWrite operation.

Reply:
The literals are illegible to see clearly, would try to understand the workflow.

Reply:
Instead of catching the raw `Exception`, should we restrict the exception to only like `File existence exception` or `File does not exist exception` ?

Reply:
> On which case we might hit this? can you help clarify please.
@nsivabalan - A task is running to create a data file.  Execution engine retried the operation on a different container/task (say stage attempt 1, due to a stage failure), while the previous task is still running in the background (not responsive to the driver).

If the second task (retry attempt) were to finish the data write and create completion marker, at the time of closing the writeHandle the first Task (in the example), would cleanup the fail after noticing the presence of the completion marker.

If the closing of the writeHandle were to happen after finalizeWrite/reconcileMarkers step, we would end up with a duplicate file, as data files created both tries will be present (will not be removed by reconcile step).

@danny0405 - Updated to HoodieException